### PR TITLE
ENH/TST: Refactor refguide-check, take 3

### PR DESCRIPTION
**Pull request by ev-br**
#### Reference issue

Continues and closes gh-16391, supersedes and closes gh-19242

#### What does this implement/fix?

Continue gh-19242 by @Sheila-nk : refactor the doctesting part of the refguide-check into a standalone tool, use pytest for running and orchestration. 

So this PR has essentially two parts: 
- heavy lifting done by Sheila (see commit attributions) 
- additional fixes to docstring examples from gh-19242. A part with pytest ignores was suggested by Lucas in https://github.com/scipy/scipy/pull/19242#discussion_r1431636535, also commit-attributed.

The remaining TODOs follow https://github.com/scipy/scipy/pull/19242#issuecomment-1719964303. What's relevant now is

- [ ] finish plumbing the external tool as git submodule
- [x] activate checking the code in tutorials
- [ ] drop the obsoleted part of `refguide-check`. I'd prefer to do it in a quick follow-up PR.

On the 'scpdt' tool side, we need to

- [x] get a better name (doctest_sp, doctest_scipy, scipy_doctest, something else?) --- `doctest-scipy` unless there are other suggestions
- [x] get out of my GH --- will propose to move it to the scipy GH org on the ML

The tool itself is in a bit of a flux at the moment, nothing blocking though. Here's the tracker which is being actively worked on: https://github.com/ev-br/scpdt/issues

The pytest plugin is written by @Sheila-nk with inputs from @melissawm and myself. Thank you Sheila and Melissa!

#### Additional information

I am starting to dislike tagging this all as _doctesting_. The goal is not ---never was and will never be --- to mix doctests into unit testing. The goal is to keep the examples in the docs current. If somebody comes up with a catchy name for this, I'm all ears :-).

### CODE REVIEW

File: scipy/optimize/_milp.py
Comment (ev-br) [inline scipy/optimize/_milp.py:325]: @mdhaber git blame points at you, so I'd ask about this change. Without it, running the example from the `milp` docstring generates a warning from numpy; adding a dtype above changes the result from `[1.0, 2.0]` to `[2.0, 2.0]`:

```
In [1]: import numpy as np

In [2]: c = -np.array([0, 1])

In [3]: A = np.array([[-1, 1], [3, 2], [2, 3]])
   ...: >>> b_u = np.array([1, 12, 12])
   ...: >>> b_l = np.full_like(b_u, -np.inf)
/home/br/mambaforge/envs/scipy-dev/lib/python3.10/site-packages/numpy/core/numeric.py:407: RuntimeWarning: invalid value encountered in cast
  multiarray.copyto(res, fill_value, casting='unsafe')

In [4]: from scipy.optimize import LinearConstraint
   ...: >>> constraints = LinearConstraint(A, b_l, b_u)

In [5]: integrality = np.ones_like(c)

In [6]: from scipy.optimize import milp
   ...: >>> res = milp(c=c, constraints=constraints, integrality=integrality)
   ...: >>> res.x
Out[6]: array([1., 2.])
```

File: scipy/sparse/__init__.py
Comment (ev-br) [inline scipy/sparse/__init__.py:233]: @dschult had to remove this line because otherwise it fails and compains about 1D slices.

File: scipy/stats/sampling.py
Comment (ev-br) [inline scipy/stats/sampling.py:73]: @chrisb83 git blame points at you: any reason `scipy.stats.sampling` did not have an `__all__` list, and would you object if it gets one?

**Review by ev-br — COMMENTED** [review]
Let me flag several tweaks to the docstrings. The fixes are mainly mechanical, may very well be completely wrong, so would appreciate a sanity check from subject experts.

**Comment by ev-br** [comment]
Testing `dev.py` options:

- [ ] `$ python dev.py test --doctests -j 4` seems to give all workers all tests instead of distributing them? Needs checking, can be done in a follow-up (`refguide-check` never supported this)
- [ ]  `$ python dev.py test --doctests -t path/to/file` does not work. Can be done in a follow-up, as `refguide-check` never supported this.

One other bikeshed item is the name of the switch: `dev.py test --doctests` (current state) or `dev.py test --doctests-only`?

With regular pytest `pytest --doctests` means run _both_ unit tests _and_ doctests; Here `dev.py test --doctests` skips regular unit tests entirely. OTOH `--doctests-only` is more typing. Thoughts?

File: scipy/stats/qmc.py
Comment (tupui) [reply scipy/stats/qmc.py:236]: Instead of `*`? Isn't it redundant then? But sure no objections otherwise 😃 Whatever works as long we don't change the imports and I can still do `sp.stats.qmc.Sobol`

**Review by tupui — COMMENTED** [review]
Really cool 🚀 

> I am starting to dislike tagging this all as doctesting. The goal is not ---never was and will never be --- to mix doctests into unit testing. The goal is to keep the examples in the docs current. If somebody comes up with a catchy name for this, I'm all ears :-).

We have smoke-tests, what about `smoke-doc`? 😅

File: scipy/stats/qmc.py
Comment (lucascolley) [reply scipy/stats/qmc.py:236]: what is the error without this change? Pretty sure this isn't the ideal solution.

File: scipy/_lib/_testutils.py
Comment (lucascolley) [inline scipy/_lib/_testutils.py:83]: can we get rid of the errors from the files which aren't submodules / vendored?

Comment (ev-br) [reply scipy/_lib/_testutils.py:83]: Probably. Together with https://github.com/ev-br/scpdt/issues/94
I'd defer it to a follow-up though.

Comment (ev-br) [reply scipy/_lib/_testutils.py:83]: That said, if you feel like tackling it in this PR Lucas, I won't mind at all :-).

Comment (lucascolley) [reply scipy/_lib/_testutils.py:83]: don't tempt me... (no promises but maybe)

File: scipy/stats/sampling.py
Comment (chrisb83) [reply scipy/stats/sampling.py:73]: I think I followed `stats/qmc.py`. I don't see a problem with the addition of __all__

**Comment by ev-br** [comment]
Here's an update:

- Following up on the ML discussion and Pamphile's suggestion (thanks @tupui), the dev.py plumbing is through a dedicated command: 
 `$ python dev.py smoke-docs`. The command mirrors the `test` command, and accepts much the same arguments: `smoke-docs -s submodule`, `smoke-docs -t path/to/file::Klass::method` etc.

- To smoke-test the tutorial, there's a separate command `$ python dev.py smoke-tutorials`

- smoke-testing the tutorials now includes subfolders--- those in `doc/source/tutorial/{stats, interpolate}` have never been checked by refguide-check.  This PR now includes several easy fixes and xfails for less obvious errors.

File: doc/source/tutorial/stats/sampling_pinv.rst
Comment (ev-br) [inline doc/source/tutorial/stats/sampling_pinv.rst:185]: @tirthasheshpatel git blame point to you in these sampling_*rst files. Could you take a look over my tweaks to see if they mask real issues?

File: scipy/conftest.py
Comment (ev-br) [inline scipy/conftest.py:359]: @tirthasheshpatel had to xfail the whole tutorial, looks like the code has changed after the docs were written?

**Review by ev-br — COMMENTED** [review]
Flagging a couple of possible issues in the tutorials.

Fixing those would be an enhancement, since these were never refguide-checked. So fixes may equally well come in this PR (send me a PR or let me know what to cherry-pick) or as follow-ups.

**Comment by lucascolley** [comment]
CI looks happy 👀

**Comment by ev-br** [comment]
Thanks Lucas. Now that CI is green, the plugin repository has been moved to https://github.com/scipy/scipy_doctest, as discussed on the mailing list. Could you check that you've a green button on that repository please?

**Comment by tupui** [comment]
Should this repo also be name smoke-docs?

**Comment by ev-br** [comment]
ATM the repository and the package are `scipy_doctest` and the dev.py stanzas are `$ python dev.py smoke-docs -s submodule` and `$ python dev.py smoke-tutorial`. The former supports the usual `-s`, `-t`, `-v`, `-- -x` knobs etc. Good enough?

**Comment by tupui** [comment]
Ah ok 👍 yes all good

**Comment by lucascolley** [comment]
Very minor but is there a reason that "smoke-docs" is plural while "smoke-tutorial" is not?

**Comment by lucascolley** [comment]
> Could you check that you've the green button on that repository please?

Doesn't look like it.image

**Comment by ev-br** [comment]
> Doesn't look like it.

Hmmm. That's not right. Somehow the 'settings' tab is missing after transferring the repo, both in the scipy org and in my personal GH account. A workaround from 2020, https://github.com/orgs/community/discussions/22222, does not seem to work for me either, the `/settings` endpoint 404s for me.

Let's give a bit in a hope there a database to sync or something.

OTOH, I now see that the 'settings' tab is missing on some repos in the scipy org for me, but not all. Which suggests there is some action to be taken by somebody with the admin-level access. I'll follow up.

**Comment by rgommers** [comment]
Should be fixed now, it was a matter of (1) adding the `maintainers` team and (2) upgrading @ev-br's rights on that repo to `admin` (not sure why that's not the default after transferring a repo). I also enabled branch protection for `main`.

**Comment by lucascolley** [comment]
Yep I have permissions now 👍

**Comment by ev-br** [comment]
Thank you Ralf for sorting this out!

**Comment by ev-br** [comment]
> Very minor but is there a reason that "smoke-docs" is plural while "smoke-tutorial" is not?

Good point! `smoke-tutorials` it is, updated.

**Comment by ev-br** [comment]
I'm planning to merge this in about a week unless further comments.

**Review by tupui — APPROVED** [review]
Go go go 👍

**Pull request merged**
